test/gke: refactor test-clusters operations

nbusseneau opened the pull request:
See individual commits.

nebril commented on test/gke/select-cluster.sh:
```diff
@@ -36,7 +36,7 @@ while [ $locked -ne 0 ]; do
 	echo "checking whether cluster ${cluster_uri} has a running operation"
-	if gcloud container operations list | grep RUNNING | grep "${cluster_name}" ; then
+	if $(gcloud container operations list --project "${project}" --filter="status=RUNNING AND targetLink=${cluster_uri}") ; then
 		echo "cluster has an ongoing operation, trying out another cluster"
```
This line will always run, `gcloud` doesn't return error on empty list:

```
$ if $(gcloud container operations list --project "cilium-ci" --filter="status=RUNNING AND targetLink=trolo"); then echo "true"; fi
Listed 0 items.
true

```

nbusseneau replied:
Yea I did see that xD Fixed in latest commit.

nebril replied:
I don't think that's fixed, I tested this locally:

```
$ if $(gcloud container operations list --project "cilium-ci" --filter="status=RUNNING AND targetLink=trolo" --format="value(name)"); then echo "true"; fi
true
```

nbusseneau replied:
Right, cause I can't Bash. Fixed (maybe).

nbusseneau committed:
test/gke: remove obsolete script

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

nbusseneau committed:
test/gke: refactor gcloud container operations

Using server-side filtering greatly accelerates checking for running
operations.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

nbusseneau committed:
test/gke: handle operations via ConfigConnector only

ConfigConnector already takes care of resizing/deletion for us, no need for manually handling that.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

nebril commented on test/gke/resize-cluster.sh:
```diff
@@ -28,22 +28,14 @@ script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
-	echo "trying to scale cluster up"
-	if gcloud container clusters resize --project "${project}" --region "${region}" --node-pool "${node_pools[0]}" --num-nodes "${node_count}" --quiet "${cluster_uri}" ; then
+	echo "waiting for resize operations to finish"
+	operations=$(gcloud container operations list --project "${project}" --filter="status=RUNNING AND operationType=UPDATE_CLUSTER AND targetLink=${cluster_uri}" --format="value(name)")
```
afaiu this will get operations list and if there are no running operations on the cluster, it will think the cluster is scaled? I see two issues here:

- the operation hasn't kicked in yet (this is why this whole setup was working in the first place, there was a delay in operation creation from configconnector which will cause this script to report scaling as successful when it has not yet started
- as is, this will check for operation four times in 15s intervals and finish the script. I think that 60s is not enough time for scale operation to finish, so we need to either change `while [ $resize_retries -lt 4 ]; do` to higher retry count or come up with an altogether different way of checking whether cluster is scaled.

nbusseneau replied:
- From my testing this was instantaneous 🤔 But yea that's a good point.
- I agree, here's a smarter solution that's not just "patching up the existing code".

qmonnet reviewed (approved):
Looks good from my side, but obviously deferring to Maciej on the cluster scaling item.

nbusseneau committed:
test/gke: fix retry loop failing

From `bash` manpage:

> ((expression))
> The expression is evaluated according to the rules described below
> under ARITHMETIC EVALUATION.  If the value of the expression is
> non-zero, the return status is 0; otherwise the return status is 1.
> This is exactly equivalent to let "expression".

With `set -o errexit`, this means `((resize_wait_retries++))` fails...

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>
